### PR TITLE
Select profiles refactor 

### PR DESCRIPTION
**Pull request by lovelysandlonelys**

### CODE REVIEW

File: oneargopy/Settings.py [hunk 1]
```diff
@@ -31,7 +31,7 @@ class DownloadSettings():
         :param: user_settings : str - An optional path to user defined 
             settings.
 
-        Settings: 
+        SETTINGS:
```
Comment (hfrenzel-noaa) [inline oneargopy/Settings.py]: WHY DO YOU USE ALL CAPS HERE???

Comment (lovelysandlonelys) [reply oneargopy/Settings.py]: I am, not sure why I did that?? I'll switch it back to lowercase so we're not yelling at anyone haha.

File: oneargopy/Argo.py [hunk 2]
```diff
@@ -18,55 +18,183 @@
...
-        if self.download_settings.verbose: print(f'\nDownloading index files...\n')
+        if self.download_settings.verbose: print(f'\nDownloading index files...')
         for file in self.download_settings.index_files:
             self.__download_index_file(file)
 
-        # Load the argo_synthetic-profile_index.txt file into a dataframe
-        if self.download_settings.verbose: print(f'\nTransferring index files into dataframes...')
-        self.sprof_index  = self.__load_sprof_dataframe()
+        # Load the index files into dataframes
+        if self.download_settings.verbose: print(f'\nTransferring index files into data frames...')
```
Comment (hfrenzel-noaa) [inline oneargopy/Argo.py]: let's keep it one word "dataframe" throughout

File: oneargopy/Argo.py [hunk 3]
```diff
@@ -227,54 +345,432 @@ def __load_sprof_dataframe(self) -> pd:
...
+        """ A function to mark weather the floats listed in prof_index are 
+            biogeochemical floats or not. 
+        """
+        bgc_floats = self.sprof_index['wmoid'].unique()
+        is_bgc = self.prof_index['wmoid'].isin(bgc_floats)
+        self.prof_index.insert(1, "is_bgc", is_bgc)
+
+    def __load_is_bgc_index(self)-> pd:
+        """ Function to create a dataframe with float IDs and
+            their is_bcg status for use in select_profiles().
```
Comment (hfrenzel-noaa) [inline oneargopy/Argo.py]: is_bgc

File: oneargopy/Argo.py [hunk 4]
```diff
@@ -227,54 +345,432 @@ def __load_sprof_dataframe(self) -> pd:
...
         
         # Splitting up parts of the first column
         dacs = prof_index['file'].str.split('/').str[0]
         prof_index.insert(0, "dacs", dacs)
 
-        wmoid = prof_index['file'].str.split('/').str[1]
+        wmoid = prof_index['file'].str.split('/').str[1].astype('int')
         prof_index.insert(1, "wmoid", wmoid)
 
         R_file = prof_index['file'].str.split('/').str[3].str.startswith('R')
```
Comment (hfrenzel-noaa) [inline oneargopy/Argo.py]: Let's switch this to D_file if it starts with 'D'

File: oneargopy/Argo.py [hunk 5]
```diff
@@ -227,54 +345,432 @@ def __load_sprof_dataframe(self) -> pd:
...
+        # Add profile_index column
+        prof_index.sort_values(by=['wmoid', 'date'], inplace=True)
+        prof_index.insert(0, "profile_index", 0)
+        prof_index['profile_index'] = prof_index.groupby('wmoid')['date'].cumcount() + 1
+
         return prof_index
+    
+
+    def __mark_bgcs_in_prof(self):
+        """ A function to mark weather the floats listed in prof_index are
```
Comment (hfrenzel-noaa) [inline oneargopy/Argo.py]: whether

File: oneargopy/Argo.py [hunk 6]
```diff
@@ -227,54 +345,432 @@ def __load_sprof_dataframe(self) -> pd:
...
+        if self.download_settings.verbose: print(f'Validating longitude and latitude limits...')
+
+        # Validating Lists
+        if len(self.lon_lim) != len(self.lat_lim):
+            raise Exception(f'The length of the longitude and latitude lists must be equal.')
+        if len(self.lon_lim) == 2:
+            if (self.lon_lim[1] <= self.lon_lim[0]) or (self.lat_lim[1] <= self.lat_lim[0]):
+                if self.download_settings.verbose: print(f'Longitude Limits: min={self.lon_lim[0]} max={self.lon_lim[1]}')
+                if self.download_settings.verbose: print(f'Latitude Limits: min={self.lat_lim[0]} max={self.lat_lim[1]}')
+                raise Exception(f'When passing longitude and latitude lists using the [min, max] format the max value must be greater than the min value.')
```
Comment (hfrenzel-noaa) [inline oneargopy/Argo.py]: format, the max value

File: oneargopy/Argo.py [hunk 7]
```diff
@@ -227,54 +345,432 @@ def __load_sprof_dataframe(self) -> pd:
...
+            if (abs(self.lon_lim[1]) - self.lon_lim[0] < self.epsilon) and (abs(self.lat_lim[1]) - self.lat_lim[0] < self.epsilon):
```
Comment (hfrenzel-noaa) [inline oneargopy/Argo.py]: if (abs(self.lon_lim[1] - self.lon_lim[0] - 360.0) < self.epsilon) and (abs(self.lat_lim[1] - self.lat_lim[0] - 180.0) < self.epsilon):

File: oneargopy/Argo.py [hunk 8]
```diff
@@ -227,54 +345,432 @@ def __load_sprof_dataframe(self) -> pd:
...
+        if not all(-90 <= lat <= 90 for lat in self.lat_lim):
+            print(f'Latitudes: {self.lat_lim}')
+            raise Exception(f'Latitude values should be between -90 and 90.')
+        
+        # Validate Longitudes
+        ## Checking range of longitude values
+        lon_range = max(self.lon_lim) - min(self.lon_lim)
+        if lon_range > 360 or lon_range <= 0:
+            if self.download_settings.verbose: print(f'Current longitude range: {lon_range}')
+            raise Exception(f'The range between the maximum and minimum longitude values must be between 1 and 360.')
```
Comment (hfrenzel-noaa) [inline oneargopy/Argo.py]: must be between 0 and 360

(0.2 or so is allowed)

File: oneargopy/Argo.py [hunk 9]
```diff
@@ -18,55 +18,183 @@
...
+        if self.download_settings.verbose: print(f'Initialize is finished!\n\n')
+
+        if not self.download_settings.keep_index_in_memory:
+            if self.download_settings.verbose: print('Removing dataframes from memory...')
+            del self.sprof_index 
+            del self.prof_index
+            del self.float_is_bgc_index
+
+
+    def select_profiles(self, lon_lim: list = [-180, 180], lat_lim: list = [-90, 90], start_date: str = '1995-01-01', end_date: str = None, **kwargs)-> dict:
```
Comment (hfrenzel-noaa) [inline oneargopy/Argo.py]: break line so that it fits into 80 columns

File: oneargopy/Argo.py [hunk 10]
```diff
@@ -18,55 +18,183 @@
...
+            NOTE: 
+            The longitude and latitude limits can be entered as either
+            two element lists, in which case the limits will be interpreted
+            as maximum and minimum limits tht form a rectangle, or they
+            can be entered as a longer list in which case each pair of longitude
+            and latitude values correspond to a vertices of a polygon.
+            The longitude and latitude limits can be input in any 360 degree
+            range that encloses all the desired longitude values.
+
+            KEY ARGUMENT VALUE OPTIONS IN PROGRESS:
```
Comment (hfrenzel-noaa) [inline oneargopy/Argo.py]: Key/argument value options in progress:

File: oneargopy/Argo.py [hunk 11]
```diff
@@ -183,39 +307,33 @@ def __try_download(self, file_name: str, update_status: bool)-> None:
...
         """
         file_name = "argo_synthetic-profile_index.txt"
+        # The header is 8 here because there are 8 lines in both index files devoted to header information.
         file_path = Path.joinpath(self.download_settings.base_dir, 'Index', file_name)
         sprof_index  = pd.read_csv(file_path, delimiter=',', header=8, parse_dates=['date','date_update'], 
                                 date_format='%Y%m%d%H%M%S')
         
-        # Parsing out variables in first column file
+        # Parsing out variables in first column: file
         dacs = sprof_index ['file'].str.split('/').str[0]
```
Comment (hfrenzel-noaa) [inline oneargopy/Argo.py]: remove blank after sprof_index

File: oneargopy/Argo.py [hunk 12]
```diff
@@ -183,39 +307,33 @@ def __try_download(self, file_name: str, update_status: bool)-> None:
...
-        sprof_index .insert(1, "wmoid", wmoid)
+        wmoid = sprof_index ['file'].str.split('/').str[1].astype('int')
+        sprof_index.insert(0, "wmoid", wmoid)
 
         profile = sprof_index ['file'].str.split('_').str[1].str.replace('.nc', '')
-        sprof_index .insert(2, "profile", profile)
+        sprof_index.insert(2, "profile", profile)
 
         # Splitting the parameters into their own columns
         parameters_split = sprof_index ['parameters'].str.split()
```
Comment (hfrenzel-noaa) [inline oneargopy/Argo.py]: remove blank after sprof_index here and in next line

File: oneargopy/Argo.py [hunk 13]
```diff
@@ -183,39 +307,33 @@ def __try_download(self, file_name: str, update_status: bool)-> None:
...
         data_types_split = sprof_index ['parameter_data_mode'].apply(list)
 
+        # R: raw data, A: adjusted mode (real-time adjusted), D: delayed mode quality controlled
         data_type_mapping = {np.nan: 0, 'R':1, 'A':2, 'D':3 }
         mapped_data_types_split = data_types_split.apply(lambda lst: [data_type_mapping.get(x, 0) if pd.notna(x) else 0 for x in lst])
```
Comment (hfrenzel-noaa) [inline oneargopy/Argo.py:443]: here and a little later, line breaks to keep under 80 chars per line

File: oneargopy/Argo.py [hunk 14]
```diff
@@ -183,39 +307,33 @@ def __try_download(self, file_name: str, update_status: bool)-> None:
...
 
-        # Create a new DataFrame from the split lists
+        # Create a new DataFrame from the split parameters 
         expanded_df = pd.DataFrame({
             'index': sprof_index .index.repeat(parameters_split.str.len()),
```
Comment (hfrenzel-noaa) [inline oneargopy/Argo.py]: remove blank after sprof_index

File: oneargopy/Argo.py [hunk 15]
```diff
@@ -227,54 +345,432 @@ def __load_sprof_dataframe(self) -> pd:
             # with pd.option_context('future.no_silent_downcasting', True):
         result_df = expanded_df.pivot(index='index', columns='parameter', values='data_type').fillna(0).astype('int8')
 
-        # Fill in parameters and dacs before removing rows
-        # Fill in source_settings information based off of synthetic file
+        # Fill in source_settings information based off of synthetic file before removing rows
```
Comment (hfrenzel-noaa) [inline oneargopy/Argo.py]: based off of sprof index file

File: oneargopy/Argo.py [hunk 16]
```diff
@@ -227,54 +345,432 @@ def __load_sprof_dataframe(self) -> pd:
...
         if self.download_settings.verbose: print(f'Filling in source settings information...')
         self.source_settings.set_avail_vars(sprof_index )
```
Comment (hfrenzel-noaa) [inline oneargopy/Argo.py]: it looks like you added blanks everywhere after sprof_index; please remove them all

File: oneargopy/Argo.py [hunk 17]
```diff
@@ -227,54 +345,432 @@ def __load_sprof_dataframe(self) -> pd:
...
         self.source_settings.set_dacs(sprof_index )
```
Comment (hfrenzel-noaa) [inline oneargopy/Argo.py]: this should use prof_index instead (not all DACs handle BGC floats)

File: oneargopy/Argo.py [hunk 18]
```diff
@@ -227,54 +345,432 @@ def __load_sprof_dataframe(self) -> pd:
...
+        """ A function to validate the value of the optional 'type' keyword argument.
+        """
+        if self.download_settings.verbose: print(f"Validating 'type' keyword argument...")
+
+        if self.float_type != 'all' and self.float_type != 'phys' and self.float_type != 'bgc':
+                raise Exception(f"The only acceptable values for the 'type' keyword argument are 'all', 'phys', and 'bgc'.")
+        
+    
+    def __validate_floats_kwarg(self):
+        """ A function to validated the 'floats' keyword argument. The 'floats' must be a list even if it is a single float.
```
Comment (hfrenzel-noaa) [inline oneargopy/Argo.py]: validate; also line lengths! Here and in previous functions

File: oneargopy/Argo.py [hunk 19]
```diff
@@ -227,54 +345,432 @@ def __load_sprof_dataframe(self) -> pd:
...
+                raise Exception(f"The only acceptable values for the 'ocean' keyword argument are 'A' (Atlantic), 'P' (Pacific), and 'I' (Indian).")
+
+
+    def __prepare_selection(self):
+        """ A function that determines what dataframes will be loaded/used when selecting floats.
+            We determine what dataframes to load based on two factors: type and passed floats.  
+
+            If type is 'phys', the dataframe based on ar_index_global_prof.txt will be used.
+            If type is 'bgc', the dataframe based on argo_synthetic-profile_index.txt will be used.
+            If type is 'all', both dataframes are used.
```
Comment (hfrenzel-noaa) [inline oneargopy/Argo.py:660]: ... are used. BGC floats are taken from argo_synthetic-profile_index.txt, non-BGC floats from ar_index_global_prof.txt.

File: oneargopy/Argo.py [hunk 20]
```diff
@@ -227,54 +345,432 @@ def __load_sprof_dataframe(self) -> pd:
...
+
+            If the user passed floats, we only load the passed floats into the selection frames.
+
+            If keep_index_in_memory is set to false the dataframes created during Argo's
```
Comment (hfrenzel-noaa) [inline oneargopy/Argo.py]: to false, the dataframes

File: oneargopy/Argo.py [hunk 21]
```diff
@@ -227,54 +345,432 @@ def __load_sprof_dataframe(self) -> pd:
...
+                  # Make a list of phys floats that the user wants 
+                  phys_filter = (self.float_is_bgc_index['wmoid'].isin(self.floats)) & (self.float_is_bgc_index['is_bgc'] == False)
+                  selected_floats_phys = self.float_is_bgc_index[phys_filter]['wmoid'].tolist()
+                  # Gather phys profiles for these floats from prof index frame
+                  self.selected_from_prof_index = self.prof_index[self.prof_index['wmoid'].isin(selected_floats_phys)]
+
+        if self.download_settings.verbose:
+            num_unique_floats = len(self.selected_from_sprof_index['wmoid'].unique()) + len(self.selected_from_prof_index['wmoid'].unique())
+            print(f"We will filter through {num_unique_floats} floats!") 
+            print(f'There are {len(self.selected_from_sprof_index) + len(self.selected_from_prof_index)} profiles are associated with these floats!')
```
Comment (hfrenzel-noaa) [inline oneargopy/Argo.py]: profiles associated with
also add line break

File: oneargopy/Argo.py [hunk 22]
```diff
@@ -227,54 +345,432 @@ def __load_sprof_dataframe(self) -> pd:
...
+    def __narrow_profiles_by_criteria(self)-> dict:
+        """ A function to narrow down the available profiles to only those
+            that meet the criteria passed to select_profiles.
+
+            :return: narrowed_profiles : dict - A dictionary with float ID
+                keys corresponding to a list of profiles that match criteria.
+        """
+        # Filter by time, space, and type constraints first.
+        if self.float_type == 'bgc' : 
+            self.selection_frame_phys = pd.DataFrame()
```
Comment (hfrenzel-noaa) [inline oneargopy/Argo.py:725]: add comment: "# empty df for concat"

File: oneargopy/Argo.py [hunk 23]
```diff
@@ -227,54 +345,432 @@ def __load_sprof_dataframe(self) -> pd:
...
+            self.selection_frame_bgc = self.__get_in_time_and_space_constraints(self.selected_from_sprof_index)
+        
+        # Set the selection frame
+        self.selection_frame = pd.concat([self.selection_frame_bgc, self.selection_frame_phys])
+
+        # Remove extraneous frames if keep index in memory is set to false
+        if not self.download_settings.keep_index_in_memory: 
+            del self.sprof_index
+            del self.prof_index
+            del self.selection_frame_bgc
```
Comment (hfrenzel-noaa) [inline oneargopy/Argo.py]: selection_frame_bgc and selection_frame_phys can always be deleted after concat, right?
so comment should be just "# Remove extraneous frames"

File: oneargopy/Argo.py [hunk 24]
```diff
@@ -227,54 +345,432 @@ def __load_sprof_dataframe(self) -> pd:
...
+        # Create polygon or box using lat_lim and lon_lim 
+        if self.download_settings.verbose: print(f'Creating polygon...')
+        if len(self.lat_lim) == 2:
+            shape = [[max(self.lon_lim), min(self.lat_lim)], # Top-right
+                     [max(self.lon_lim), max(self.lat_lim)], # Bottom-right
+                     [min(self.lon_lim), max(self.lat_lim)], # Bottom-left
+                     [min(self.lon_lim), min(self.lat_lim)]] # Top-left
+        else:
+            shape = []
+            for lat, lon in zip(self.lat_lim, self.lon_lim):
```
Comment (hfrenzel-noaa) [inline oneargopy/Argo.py]: reverse lat,lon both times to match "shape"

File: oneargopy/Argo.py [hunk 25]
```diff
@@ -227,54 +345,432 @@ def __load_sprof_dataframe(self) -> pd:
...
+        if self.download_settings.verbose: 
+            profiles_in_range_dataframe = dataframe_to_filter[profiles_in_range]
+            print(f"{len(profiles_in_range_dataframe['wmoid'].unique())} floats fall within the geographic range!")   
+            print(f'{len(profiles_in_range_dataframe)} profiles associated with those floats!')
+
+        return profiles_in_range
+
+
+    def __get_in_date_range(self, dataframe_to_filter: pd)-> list:
+        """ A function to create and return two true false arrays indicating
```
Comment (hfrenzel-noaa) [inline oneargopy/Argo.py]: True/False arrays
but it's really just one, isn't t?

File: oneargopy/Argo.py [hunk 26]
```diff
@@ -227,54 +345,432 @@ def __load_sprof_dataframe(self) -> pd:
...
+        # Printing Dict, likely to remove after testing period
+        print(f'Floats: {selected_floats_dict.keys()}')
+        for key, value in selected_floats_dict.items():
+            print(f'{key}: {value}')
+
+        return selected_floats_dict
+        
+
+    def __get_in_geographic_range(self, dataframe_to_filter: pd)-> list:
+        """ A function to create and return two true false arrays indicating
```
Comment (hfrenzel-noaa) [inline oneargopy/Argo.py]: see below for time range
just one True/False array, right?

File: oneargopy/Argo.py (hunk 23 above)
Comment (lovelysandlonelys) [reply oneargopy/Argo.py]: image

File: oneargopy/Argo.py (hunk 24 above)
Comment (lovelysandlonelys) [reply oneargopy/Argo.py]: image

File: oneargopy/Argo.py (hunk 25 above)
Comment (lovelysandlonelys) [reply oneargopy/Argo.py]: ah yes, i forgot to change the description after updating that method

File: oneargopy/Argo.py [hunk 27]
```diff
@@ -183,98 +419,494 @@ def __try_download(self, file_name: str, update_status: bool)-> None:
...
+            The 'floats' must be a list even if it is a single float.
+        """
+        # Casting to list
+        if not isinstance(self.floats, list):
+            self.floats = [int(self.floats)]
+
+        # Checking that floats are in index dataframes
+        all_present = all(float in self.prof_index['wmoid'].values for float in self.floats)
+        if not all_present: 
+            raise Exception(f"You have passed float values that do not exist in the dataframes: {self.floats}")
```
Comment (hfrenzel-noaa) [inline oneargopy/Argo.py]: float IDs that

Comment (hfrenzel-noaa) [inline oneargopy/Argo.py]: Can you show only the float IDs that are not present in the index? (easy test: pass in 1,4903500; 1 should be listed)

File: oneargopy/Argo.py [hunk 28]
```diff
@@ -18,56 +18,296 @@
...
         if self.download_settings.verbose: self.__display_floats() 
 
-        if self.download_settings.verbose: print(f'Initialize is finished!')
+        if self.download_settings.verbose: print(f'Initialize is finished!\n\n')
+
+        if not self.download_settings.keep_index_in_memory:
+            if self.download_settings.verbose: print('Removing dataframes from memory...')
+            del self.sprof_index 
+            del self.prof_index
+            del self.float_is_bgc_index
```
Comment (hfrenzel-noaa) [inline oneargopy/Argo.py]: float_is_bgc_index is much smaller (~20000 rows), so we should always keep it in memory, saves us the del/reload

File: oneargopy/Argo.py (hunk 27 above)
Comment (lovelysandlonelys) [reply oneargopy/Argo.py]: image

image

**Review by hfrenzel-noaa — APPROVED** [review]

**Pull request merged**